### ISSUE TITLE
Create specific types for each NodeList

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#1205]
I've brought this up a few times before and it didn't get much enthusiasm. I don't really know why, so here's a discussion issue.

Currently we have `NodeList<SomeNodeType>`.
Things we are not able to do with this:
- Visit them cleanly (#1056)
- Ask what type is in them (#850)

If we added a field for the type that is in them, constructing nodelists would become a pain:
`new NodeList<>(ImportDeclaration.class)`

If we created subclasses for every type, well, that solves everything:
```java
ImportDeclarationList list = new ImportDeclarationList();
list.getType(); // -> ImportDeclaration.class
```

```java
interface XxxVisitor {
   void visit(ImportDeclarationList l);
   void visit(NameList l);
   // ...
}
```

Every list is of course a child of `NodeList<...>`.

**Comment by leus** [comment]
Yes to specialized types, so we don't have to do ugly stuff like [this](https://github.com/javaparser/javaparser/commit/095128cebabf91e1676a952ab966944ef475e173#diff-4525adfd97dacd0f0f8cfa9468aa6f2eR1396).
